### ISSUE TITLE
How can I run some logic only the first time something becomes visible?

### ISSUE THREAD

**Issue by tremby** [thebuilder/react-intersection-observer#233]
I realize there is the `triggerOnce` option which I can set to true.

But how in my component can I run some logic, such as send a tracking event, the first time the element comes in to view? I can't just send it if `inView` is true, since if the component renders again, it'll still be true and it'll happen again.

Must I keep track of whether I've sent it already in state or similar?

Or is there some way I'm not thinking of?

**Comment by thebuilder** [comment]
For your usecase I would combine `triggerOnce` with the `onChange` callback. The callback will only be called once, and can then fire any tracking events you need.

Something like: 
```
const Component = () => (
  <InView as="div" triggerOnce onChange={inView => dataLayer.push('Element visible')}/>
)
```

**Comment by thebuilder** [comment]
I wrote a few more details about this:
https://github.com/thebuilder/react-intersection-observer/blob/master/docs/Recipes.md#track-impressions

**Comment by tremby** [comment]
Thanks for this. Your example wouldn't fire again if for example the component rerenders, or its parent does?

**Comment by thebuilder** [comment]
No. If you set `triggerOnce` it removes the `IntersectionObserver` after it triggers the first time. You would have to unmount the component before it triggers again.

**Comment by tremby** [comment]
I see. I was worried the useEffect would happen again, since from what I understand effect hooks run their "unmount" action and then their "mount" again each time the component rerenders, and on a rerender 'inView' would still be true. I do not pretend to understand the useEffect hook well though!

https://reactjs.org/docs/hooks-effect.html

> **Does useEffect run after every render?** Yes! By default, it runs both after the first render and after every update. (We will later talk about how to customize this.) Instead of thinking in terms of “mounting” and “updating”, you might find it easier to think that effects happen “after render”. React guarantees the DOM has been updated by the time it runs the effects.

**Comment by thebuilder** [comment]
By default, yes. But you can control when they run. The useInView hook is optimized to only run again if the props change. 

See https://github.com/thebuilder/react-intersection-observer/blob/master/src/useInView.tsx

**Comment by tremby** [comment]
Alright, I'm on a different project where I want to do something similar but a little different. This time I want to run some logic every time the thing becomes visible, so I don't want to use `triggerOnce`. If the only state in the component is what `useInView` gives me, this works fine -- if I run some code in a conditional if `inView` is true, it only happens when the element comes into view, once each time.

But if I have other state changes via `useState` etc, `useInView` is true every time the render function runs while the element is in view.

How, in the presence of other changing state, would I run some logic only when the value of inView changes one way or the other?

**Comment by tremby** [comment]
I figured it out. Adding my solution here in case anyone comes across this.

I first tried using a new state var to track what the previous value of `inView` was, but of course every time this gets updated it causes an infinite render loop. Maybe there's some way around that, maybe not.

But it works correctly if I track the previous value of `inView` with a ref instead of something in state.

```js
import { useRef } from 'react';

...

const MyComponent= () => {
  const prevInView = useRef(false)
  const [ref, inView] = useInView(...)

  if (!prevInView.current && inView) {
    console.log("Wasn't in view, now is")
  } else if (prevInView.current && !inView) {
    console.log("Was in view, now isn't")
  }
  prevInView.current = inView

  return ...
}
```

**Comment by tremby** [comment]
> By default, yes. But you can control when they run. The useInView hook is optimized to only run again if the props change.

I'm looking again and this doesn't actually help, because this does nothing to guarantee that your functional component (its render method) is only going to run when the useInView hook changes its output value. All sorts of other things can cause it to render again, like props or state changing. So the example in the readme is only valid (only guarantees that the event will only be tracked once) if this hook is the only thing which can cause it to rerender. As soon as there's other state or props you'll get the event tracked every time it rerenders. I think this could lead to a lot of weird bugs. Maybe worth mentioning in the readme, even if you think addressing it is out of scope?

**Comment by thebuilder** [comment]
This is how Hooks work - it is run everytime your component renders, a you always get the value back. If you need to then track if it changed since the last render, you would do that in your component with a `usePrevious` (like in your solution).

https://reactjs.org/docs/hooks-faq.html#how-to-get-the-previous-props-or-state

**Comment by tremby** [comment]
Yes, I understand this. My point is that the example code you give in your readme, which promises to track an event only once, will fail (track the event more times) as soon as somebody has more hooks or props which will cause their render method to run again.

**Comment by thebuilder** [comment]
The example will only fire if the boolean value of `inView` is changed.
https://reactjs.org/docs/hooks-effect.html#tip-optimizing-performance-by-skipping-effects

**Comment by tremby** [comment]
I'm not sure how but I got really mixed up. You're absolutely right; sorry for the noise.
